### DISCUSSION TITLE
Is Clickhouse good for indexing crypto blockchains?

### DISCUSSION THREAD

**Discussion by fridary** [ClickHouse/ClickHouse#72182] [Q&A]
Hello everybody.

I have some experience with ClickHouse in the past, but now I have a problem to index Ethereum blockchain to a new database. Now it's in Postgres, but it's getting slower as there are millions of transactions and rows, even using indexes some responses can reach half a minute.

So, there are some things that I am worry about and would like to ask if ClickHouse can manage this:

1) ethereum blockchain, as every blockchain, has blocks. It creates every 12 seconds new blocks and each block has 100-300 transactions, each transaction has 1-100 logs. Let's say we have blocks, transactions and logs tables.
Sometimes it can happen blocks reordering, that means I have to delete past data from a table and insert new.
Block reordering happens around every 5 hours and it takes 1 row to delete from block table, 100-300 rows delete from transactions table and can reach 10k rows to delete in logs table.
I see there is "The Lightweight DELETE Statement" in docs. If I use this method, would it complicate database and make slower next SELECT requests? Every ~5 hours I need to delete ~10k rows from a table of millions rows.

2) blocks, transactions, logs - all these tables depend on each other and errors can happen. That's why I need to use COMMIT and ROLLBACK sql statements, when inserting data consistently (in series). Here are docs:
https://clickhouse.com/docs/en/guides/developer/transactional#transactions-commit-and-rollback
But it says it's experimental. So, will it work for my problem? 

3) what family engine should I use (MergeTree, ReplacingMergeTree, AggregatingMergeTree, ...)? I need simply add data to some tables and create indexes for fast SELECT queries. Some tables contain fields with unique index. Sometimes do lightweight delete.

**Comment by UnamedRus** [comment]
> what family engine should I use (MergeTree, ReplacingMergeTree, AggregatingMergeTree, ...)? I need simply add data to some tables and create indexes for fast SELECT queries. Some tables contain fields with unique index. Sometimes do lightweight delete.

ReplacingMergeTree

> blocks, transactions, logs - all these tables depend on each other and errors can happen.
> That's why I need to use COMMIT and ROLLBACK sql statements, when inserting data consistently (in series). Here are docs:
> But it says it's experimental. So, will it work for my problem?

experimental transactions is not good fit for that, as they experimental,  but other approaches are possible.

> I see there is "The Lightweight DELETE Statement" in docs. If I use this method, would it complicate database and make slower next SELECT requests? Every ~5 hours I need to delete ~10k rows from a table of millions rows.

Most likely, it wouldn't work that well, plus you need to have tx support across tables.

> Block reordering happens around every 5 hours and it takes 1 row to delete from block table, 100-300 rows delete from transactions table and can reach 10k rows to delete in logs table.

There is pattern, which was used by other crypto companies, basically idea is to have git-like behavior.

You will have main branch, which is rarely or if even "corrected" and often changed tail (latest hour/day of data)

So your queries will look like

SELECT * FROM transactions WHERE (branch = 'main' AND block_number < 3000000) OR (block_number > 3000000 AND block_hashes IN (.....)) 

Where 
> (.....) can be passed with each query by client/stored in small table in CH/ Set table engine/Dictionary 

And every day/hour you will have ETL, which will move ` block_number < 3000000) OR (block_number > 3000000` to next mark

**Reply by fridary** [reply]
> as they experimental, but other approaches are possible.

Can you tell me please how to do that?

**Reply by UnamedRus** [reply]
TLDR, you can have 2 tables, one for data and second will be kind of transaction log, and you can do lookups in main table by it.

SELECT * FROM transactions WHERE insert_block_id IN (SELECT insert_block_id  FROM tx_log WHERE ...) OR event_date < today();

**Comment by rschu1ze** [comment]
Duplicate of https://stackoverflow.com/questions/79211889/is-clickhouse-good-for-indexing-crypto-blockchains
